### PR TITLE
[dagster-airflow] from airflow to dagster guide

### PR DESCRIPTION
**Pull request by Ramshackle-Jamathon**
This pr adds a "from airflow to dagster guide" that walks through rewriting the airflow tutorial dag into a dagster job, geared towards getting a developer familiar with airflow up to speed with dagster.

### CODE REVIEW

**Review by rexledesma — CHANGES_REQUESTED** [review]
There are a lot of moving parts here. In this PR, I will focus only on reviewing on the "From Airflow to Dagster" section.

Here's a breakdown that I think makes sense to ease review of the current stuff here that's trying to land:

1. One PR that's focused on adding the concept mapping from Airflow to Dagster. This should be in a separate page, as a sort of glossary.
2. One PR that translates the [Airflow tutorial](https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html) into Dagster land. In the spirit of gradual exposure of complexity, only the used subset of concepts in (1) should be mentioned in (2).
3. One PR that documents the migration guide from Airflow to Dagster
4. One PR that rearranges the API docs.

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+# Learning Dagster from Airflow
+
+Airflow users often come to Dagster for a variety of reasons. Some want to take advantage of our great story around testability. Others want to deploy to an orchestrator that is container native and has better support for multiple Python environments. And still others are attracted to the promise of a better abstraction: software-defined assets.
+
+In this tutorial, we'll help you make the switch from Airflow to Dagster. Map the concepts you're familiar with in Airflow to Dagster and provide some examples of the work you did in Airflow can be done in Dagster.
+
+By the end of this tutorial, you will:
+
+- Understand how concepts in Airflow map onto Dagster
+- Understand how to author a Dagster Pipeline similiar to a Scheduled Airflow DAG
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:15]: There will be a single point of this tutorial: we will define the DAG from the [Airflow tutorial](https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html) as a Dagster job.

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+---
+
+## Terminology
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:19]: Move all the concepts to a separate page. Only mention the subset of concepts that are used in this tutorial.

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+| Datasets              | [Software-Defined Assets](/concepts/assets/software-defined-assets)                                                                                                                                                                                                                             | SDAs are much more powerful and mature than datasets, include support for things like partitioning                                                                                                |
+| Connections/Variables | [Run Config](/concepts/configuration/config-schema), [Configured APIs](/concepts/configuration/configured) and [Environment Variables](https://docs.dagster.io/dagster-cloud/developing-testing/environment-variables-and-secrets#dagster-cloud-environment-variables-and-secrets) (cloud only) |                                                                                                                                                                                                   |
+
+## Comparing an Airflow Dag to Dagster
+
+In this guide we will rewrite an Airflow Dag as a Dagster [Job](/concepts/ops-jobs-graphs/jobs).
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:44]:
```suggestion
In this guide we will rewrite an [Airflow DAG](https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html) as a Dagster [Job](/concepts/ops-jobs-graphs/jobs).
```

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+2. The definition of the schedule
+3. The configuration of that graph and schedule
+
+### Defining the graph
+
+A [Job](/concepts/ops-jobs-graphs/jobs) is made up of a [Graph](/concepts/ops-jobs-graphs/graphs) of [Ops](/concepts/ops-jobs-graphs/ops). If you've used the Airflow taskflow api this will feel familiar, With ops you will be focused on writing a graph with python functions as the nodes with the data dependencies between those as edges.
+
+After rewriting the core compute of the tutorial Airflow Dag in Dagster we're left with:
+
+```python
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:131]: This should be interpolated from a Python file

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    t1 >> [t2, t3]
+```
+
+In order to rewrite this DAG in Dagster we will break it down into three parts
+
+1. The graph of compute
+2. The definition of the schedule
+3. The configuration of that graph and schedule
+
+### Defining the graph
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:125]: Let's use the steps that you mentioned above. I think we can do configuration as we go.

1. Define the computations: "the `op`s" - in Airflow, the operators
2. Define the graph: the `job` - in Airflow, the DAG
3. Define the schedule - In Airflow, the schedule (how simple!)

```suggestion
### Step 1: Define the computations
```

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+import time
+from datetime import date, timedelta
+
+from dagster import job, op
+
+@op
+def print_date():
+    dt = date.today()
+    print(dt)
+    return dt
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:141]: We should be very precise in the mapping here. Include the docstrings, retries, use `shutil` rather than `datetime` to emulate the `BashOperator`, etc.

```suggestion
@op(name="print_date")
def t1():
    # use shutil
```

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+def templated(dt: date):
+    for i in range(5):
+        print(dt)
+        print(dt - timedelta(days=7))
+
+@job
+def tutorial_job():
+    dt = print_date()
+    sleep(dt)
+    templated(dt)
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:157]: This should be mentioned in step 2: Defining the graph

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+@repository
+def rewrite_repo():
+    return [tutorial_job]
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:161]: not mentioned in the airflow tutorial, we should not expose this yet. Maybe mention this in Step 4: Running locally?

I think this can be in a separate PR. Steps 4-7 is what's covered in the Testing section in the Airflow quickstart.

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    dt = print_date()
+    sleep(dt)
+    templated(dt)
+
+schedule = ScheduleDefinition(job=tutorial_job, cron_schedule="@daily")
+
+@repository
+def rewrite_repo():
+    return [tutorial_job, schedule]
+```
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:260]: I think all of this should be done within the steps, and should not be mentioned separately. Otherwise, we are mixing the purposes of this guide with concept mapping.

**Review by rexledesma — CHANGES_REQUESTED** [review]

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx (hunk 9 above)
Comment (slopp) [reply docs/content/integrations/airflow/from-airflow-to-dagster.mdx:260]: Of these two options, I would vote to include the concepts in the tutorial, or at least to include retries. Retries are a critical part of Airflow that Prefect also makes a big deal about. The mapping to dagster concepts is not straightforward (op retries vs job retries) so it'd be nice to elaborate in the tutorial.

Comment (rexledesma) [reply docs/content/integrations/airflow/from-airflow-to-dagster.mdx]: I think to have a baseline for this tutorial, we should only focus mapping the Airflow tutorial to its Dagster equivalent.

Retries at the operator and DAG level are mentioned in the Airflow tutorial, so we should definitely mention them in our tutorial.

But my suggestion is not to go too into depth in the tutorial for what the concept is. We should just link out the Concept Mapping guide, which will explain the parity between Airflow's retries and our retries. I'm thinking at the end of each step in our tutorial, we link a "See also: `retries`, `ops`, ...". There, the user can see the in-depth comparison between the specific concepts.

Otherwise, we risk information dumping so much content onto the new user.

File: examples/docs_snippets/docs_snippets/integrations/airflow/tutorial_rewrite.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
+# start_imports
+import os
```
Comment (Ramshackle-Jamathon) [inline examples/docs_snippets/docs_snippets/integrations/airflow/tutorial_rewrite.py:2]: @rexledesma  I've got a more pure rewrite here using the bash commands. Im interested if you think this exposes too many dagster concepts? specifically the `ins` and propagating schedule info from `tags` has me concerned.

In the `tutorial_rewrite_dagster.py` file i've got a little more streamlined python-only rewrite but we might want to make an even more simplified version thats focused on being functionally equivalent? idk we might want to create our own airflow dag to do the migrating from to make it a little simpler

File: examples/docs_snippets/docs_snippets/integrations/airflow/tutorial_rewrite_dagster.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    ins={"start": In(Nothing)}
+)
+def sleep():
+    time.sleep(5)
+
+@op
+def templated(context, ds: datetime):
+    for i in range(5):
+        context.log.info(ds)
+        context.log.info(ds - timedelta(days=7))
```
Comment (Ramshackle-Jamathon) [inline examples/docs_snippets/docs_snippets/integrations/airflow/tutorial_rewrite_dagster.py:38]: if users do run the tutorial highlighting the event logs will be nice

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+
+We can now load this file with dagit, the interactive web-ui for dagster
+
+```bash
+dagit -f <your_dagster_file>.py
+```
+
+for reference here's what the file should look like
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:158]: This should be a separate section "## Completed Code Example", similar to what we have in Snowflake: https://docs.dagster.io/integrations/snowflake

**Review by rexledesma — CHANGES_REQUESTED** [review]
almost there! let's just be explicit in each step about what's happening

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+  />
+</center>
+
+##### Op Level Retries
+
+In the tutorial Dag the sleep task allowed for 3 retries, to configure this in dagster you can use [Op level retry policies](/concepts/ops-jobs-graphs/op-retries)
+
+### Step 2: Defining the Job
+
+In Dagster, the computations defined in `op`s are composed in `job`s, which define the sequence and dependency structure of the computations you want to execute. This directly corresponds to a DAG in Airflow. Here, we compose the `op`s `print_date`, `sleep` and `templated` to match the dependency structure defined by the Airflow operators `t1`, `t2`, and `t3`.\`\`\`
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:112]:
```suggestion
In Dagster, the computations defined in `op`s are composed in `job`s, which define the sequence and dependency structure of the computations you want to execute. This directly corresponds to a DAG in Airflow. Here, we compose the `op`s `print_date`, `sleep` and `templated` to match the dependency structure defined by the Airflow operators `t1`, `t2`, and `t3`.
```

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+    templated(ds)
+```
+
+##### Job Level Retries
+
+Job level retries are managed by the [run launcher](/deployment/run-retries) you will need to enable support for them in your dagster.yaml, once you do you can define the retry count on the job.
+
+### Step 3: Defining the schedule
+
+In Dagster Schedules are defined seperately from your computations, Below we define a schedule that will run the `tutorial_job` daily:
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:128]: lowercase 
```suggestion
In Dagster, schedules can be defined for jobs, which determine the cadence at which a job is triggered to be executed. Below we define a schedule that will run the `tutorial_job` daily:
```

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+
+```python file=/integrations/airflow/tutorial_rewrite_dagster.py startafter=start_schedule endbefore=end_schedule
+schedule = ScheduleDefinition(job=tutorial_job, cron_schedule="@daily")
+```
+
+### Step 4: Running locally
+
+In order to run our newly defined Dagster Job we'll need to define a Dagster [Repository](/concepts/repositories-workspaces/repositories). A Dagster Repository is a logical grouping of Jobs and Schedules that provides an entrypoint for viewing and executing Jobs.
```
Comment (rexledesma) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:136]:
```suggestion
In order to run our newly defined Dagster job we'll need to define a Dagster [repository](/concepts/repositories-workspaces/repositories). A Dagster repository is a logical grouping of jobs and schedules that provides an entrypoint for viewing and executing them.
```

**Review by rexledesma — APPROVED** [review]
last nits!

**Comment by erinkcochran87** [comment]
I was in the process of reviewing this and saw it get merged :| I can either submit my review and we address this in a subsequent PR or I'll add it to the docs backlog.

**Comment by Ramshackle-Jamathon** [comment]
@erinkcochran87 sorry! Depending on the feedback I can revert and reopen this pr or do a follow up 🙇

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 16]
```diff
@@ -0,0 +1,200 @@
+---
+title: "Learning Dagster from Airlfow"
+description: How to get started with Dagster from an Airflow background
+---
+
+# Learning Dagster from Airflow
+
+In this tutorial, we'll help you make the switch from Airflow to Dagster. Here, we review an Airflow DAG and show how the same functionality and be achieved in Dagster.
```
Comment (erinkcochran87) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:8]: This should be `and show how the same functionality can be achieved in Dagster.`

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+---
+
+## Comparing an Airflow Dag to Dagster
+
+In this guide we will rewrite an [Airflow DAG](https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html) as a Dagster [Job](/concepts/ops-jobs-graphs/jobs). Starting with a basic Airflow DAG:
```
Comment (erinkcochran87) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:14]: > In this tutorial, we'll rewrite an [Airflow DAG](https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html) as a Dagster [job](/concepts/ops-jobs-graphs/jobs). Let's start with a basic Airflow DAG:

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+    width={2200}
+    height={1300}
+  />
+</center>
+
+##### Op Level Retries
+
+In the tutorial DAG, the `t2` operator allowed for 3 retries. To configure the same behavior in Dagster, you can use [op level retry policies](/concepts/ops-jobs-graphs/op-retries)
+
+### Step 2: Defining the Job
```
Comment (erinkcochran87) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:110]: `## Step 2: Define the job`

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+
+In Dagster, the computations defined in `op`s are composed in `job`s, which define the sequence and dependency structure of the computations you want to execute. This directly corresponds to a DAG in Airflow. Here, we compose the `op`s `print_date`, `sleep` and `templated` to match the dependency structure defined by the Airflow operators `t1`, `t2`, and `t3`.
```
Comment (erinkcochran87) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:112]: > In Dagster, the computations defined in ops are composed in jobs, which define the sequence and dependency structure of the computations you want to execute. This directly corresponds to a DAG in Airflow. Here, we compose the ops `print_date`, `sleep` and `templated` to match the dependency structure defined by the Airflow operators `t1`, `t2`, and `t3`.

No need to put ops and jobs in code format.

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+
+```python file=/integrations/airflow/tutorial_rewrite_dagster.py startafter=start_job endbefore=end_job
+@job(tags={"dagster/max_retries": 1, "dag_name": "example"})
+def tutorial_job():
+    ds = print_date()
+    sleep(ds)
+    templated(ds)
+```
+
+##### Job Level Retries
```
Comment (erinkcochran87) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:122]: `### Job-level retries`

Headings should use sentence casing.

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+
+Job level retries are managed by the [run launcher](/deployment/run-retries) you will need to enable support for them in your dagster.yaml, once you do you can define the retry count on the job.
```
Comment (erinkcochran87) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:124]: > Job-level retries are managed by the [run launcher](/deployment/run-retries). Once enabled in your `dagster.yaml` file, you can define the retry count for the job.

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+
+### Step 3: Defining the schedule
```
Comment (erinkcochran87) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:126]: ` ## Step 3: Define the schedule`

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+
+In Dagster, schedules can be defined for jobs, which determine the cadence at which a job is triggered to be executed. Below we define a schedule that will run the `tutorial_job` daily:
+
+```python file=/integrations/airflow/tutorial_rewrite_dagster.py startafter=start_schedule endbefore=end_schedule
+schedule = ScheduleDefinition(job=tutorial_job, cron_schedule="@daily")
+```
+
+### Step 4: Running locally
```
Comment (erinkcochran87) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:134]: `## Step 4: Run Dagster locally`

Or something along these lines

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 24]
```diff
@@ -0,0 +1,200 @@
...
+
+In order to run our newly defined Dagster job we'll need to define a Dagster [repository](/concepts/repositories-workspaces/repositories). A Dagster repository is a logical grouping of jobs and schedules that provides an entrypoint for viewing and executing them.
```
Comment (erinkcochran87) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:136]: > To run our newly defined Dagster job, we'll need to define a Dagster [repository](/concepts/repositories-workspaces/repositories). A Dagster repository is a logical grouping of jobs and schedules that provides an entry point for viewing and executing them.

Editing for clarity. With the shift to `Definitions`, we should update this section if everything will work with `Definitions`.

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 25]
```diff
@@ -0,0 +1,200 @@
...
+
+```python file=/integrations/airflow/tutorial_rewrite_dagster.py startafter=start_repo endbefore=end_repo
+@repository
+def rewrite_repo():
+    return [tutorial_job, schedule]
+```
+
+We can now load this file with dagit, the interactive web-ui for dagster
```
Comment (erinkcochran87) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:144]: > We can now load this file with Dagit, Dagster's web-based UI:

Additionally, I think it'd be helpful to tell users where they need to run this command.

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 26]
```diff
@@ -0,0 +1,200 @@
...
+
+```bash
+dagit -f <your_dagster_file>.py
+```
+
+## Completed Code Example
```
Comment (erinkcochran87) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:150]: > Completed code example

Making the heading use sentence case

> That's it! By now, your code should look like this:

Adding copy to this section to make it (super) clear what it is.

File: docs/content/integrations/airflow/from-airflow-to-dagster.mdx [hunk 27]
```diff
@@ -0,0 +1,200 @@
...
+description: How to get started with Dagster from an Airflow background
+---
+
+# Learning Dagster from Airflow
+
+In this tutorial, we'll help you make the switch from Airflow to Dagster. Here, we review an Airflow DAG and show how the same functionality and be achieved in Dagster.
+
+---
+
+## Comparing an Airflow Dag to Dagster
```
Comment (erinkcochran87) [inline docs/content/integrations/airflow/from-airflow-to-dagster.mdx:12]: `## Comparing an Airflow DAG to Dagster`
